server/resp/src/test/java/org/infinispan/server/resp/commands/list/blocking/PollListenerSynchronizerTest.java

```diff
@@ -0,0 +1,223 @@
+   public void onMultipleEventsAndNullPollTest(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      synchronizer.onEvent("key1".getBytes());
+      synchronizer.onEvent("key2".getBytes());
+      synchronizer.onPollComplete(null);
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("key1".getBytes(), "test-val".getBytes()));
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void onMultipleEventsAndPollTest(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      synchronizer.onEvent("key1".getBytes());
+      synchronizer.onEvent("key2".getBytes());
+      synchronizer.onPollComplete(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void onEventAndPollAndEventTest(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      synchronizer.onEvent("key1".getBytes());
+      synchronizer.onPollComplete(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+      synchronizer.onEvent("key2".getBytes());
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void onEventNullPollThenEventTest(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      synchronizer.onEvent("key1".getBytes());
+      synchronizer.onPollComplete(null);
+      synchronizer.onEvent("key2".getBytes());
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("key1".getBytes(), "test-val".getBytes()));
+   }
+
+   @Test(dataProvider = "getSynchronizer")
+   public void raceEventAndPoll(BLPOP.PollListenerSynchronizer synchronizer)
+         throws InterruptedException, ExecutionException {
+      var l = new CountDownLatch(1);
+      new Thread(() -> {
+         try {
+            l.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onEvent("key1".getBytes());
+      }).start();
+      new Thread(() -> {
+         try {
+            l.await();
+         } catch (InterruptedException e) {
+            // TODO Auto-generated catch block
+         }
+         synchronizer.onPollComplete(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
+      }).start();
+      l.countDown();
+      assertThat(synchronizer.getResultFuture().get())
+            .containsExactlyElementsOf(Arrays.asList("poll-key1".getBytes(), "poll-val1".getBytes()));
```

Comment:
I guess we can't tell for sure which one completes.

Reply:
In this case synchronizer should guarantee that poll value is returned